### PR TITLE
[TACHYON-167] Use FileSystem Interface

### PR DESCRIPTION
**Pull request by calvinjia**
Updates internal code to use the FileSystem Interface.

### CODE REVIEW

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java [hunk 1]
```diff
@@ -21,80 +21,69 @@
...
+import tachyon.client.file.options.OpenFileOptions;
 import tachyon.exception.TachyonException;
-import tachyon.thrift.FileInfo;
 
 public final class TachyonFSTestUtils {
   /**
    * Creates a simple file with {@code len} bytes.
    *
-   * @param tfs a {@link TachyonFileSystem} handler
+   * @param tfs a {@link tachyon.client.file.FileSystem} handler
```
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: can be just FileSystem

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java [hunk 2]
```diff
@@ -21,80 +21,69 @@
...
+  public static void createByteFile(FileSystem tfs, String fileName, int len,
+      CreateFileOptions options) throws IOException {
+    createByteFile(tfs, new TachyonURI(fileName), options, len);
   }
 
   /**
    * Creates a simple file with {@code len} bytes.
    *
-   * @param tfs a {@link TachyonFileSystem} handler
+   * @param tfs a {@link tachyon.client.file.FileSystem} handler
```
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: ditto

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java (hunk 1 above)
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: applies throughout this file

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java [hunk 3]
```diff
@@ -111,48 +99,38 @@ public static TachyonFile createByteFile(TachyonFileSystem tfs, TachyonURI fileU
...
    * Returns a list of files at a given {@code path}.
    *
-   * @param tfs a {@link TachyonFileSystem} handler
+   * @param tfs a {@link tachyon.client.file.FileSystem} handler
    * @param path a path in tachyon file system
    * @return a list of strings representing the file names under the given path
    * @throws IOException if {@code path} does not exist or is invalid
    */
-  public static List<String> listFiles(TachyonFileSystem tfs, String path) throws IOException {
+  public static List<String> listFiles(FileSystem tfs, String path) throws IOException {
```
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: s/tfs/fs

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java [hunk 4]
```diff
@@ -21,80 +21,69 @@
...
    * @param fileName the name of the file to be created
    * @param len file size in bytes
    * @param options options to create the file with
-   * @return the {@link TachyonFile} representation of the created file
    * @throws IOException if {@code path} is invalid (e.g., illegal URI)
    */
-  public static TachyonFile createByteFile(TachyonFileSystem tfs, String fileName, int len,
-      OutStreamOptions options) throws IOException {
-    return createByteFile(tfs, new TachyonURI(fileName), options, len);
+  public static void createByteFile(FileSystem tfs, String fileName, int len,
```
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: s/tfs/fs

Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: applies throughout this file

File: clients/unshaded/src/main/java/tachyon/client/file/options/CreateFileOptions.java [hunk 5]
```diff
@@ -47,7 +47,7 @@ public static CreateFileOptions defaults() {
    * Creates a new instance with defaults from the configuration.
    */
   private CreateFileOptions() {
-    mRecursive = false;
+    mRecursive = true;
```
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/file/options/CreateFileOptions.java]: why is that?

File: clients/unshaded/src/main/java/tachyon/client/file/FileSystem.java [hunk 6]
```diff
@@ -46,7 +49,16 @@
...
  */
 @PublicApi
-interface FileSystem {
+public interface FileSystem {
+  class Factory {
+    public static FileSystem create() {
+      if (ClientContext.getConf().getBoolean(Constants.USER_LINEAGE_ENABLED)) {
+        return LineageFileSystem.get();
+      }
+      return new BaseFileSystem();
```
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/file/FileSystem.java]: why does LineageFileSystem have a get() factory and BaseFileSystem a constructor?

File: clients/unshaded/src/main/java/tachyon/client/table/AbstractTachyonRawTables.java [hunk 7]
```diff
@@ -73,18 +71,19 @@ public RawTableInfo getInfo(RawTable rawTable) throws IOException, TachyonExcept
...
     RawTableInfo info = getInfo(column.getRawTable());
     TachyonURI columnUri = new TachyonURI(PathUtils.concatPath(info.getPath(),
         Constants.MASTER_COLUMN_FILE_PREFIX + column.getColumnIndex()));
-    TachyonFile file = mTachyonFileSystem.open(columnUri);
-    return mTachyonFileSystem.listStatus(file).size();
+    return mFileSystem.listStatus(columnUri).size();
+  }
+
+  @Override
+  public TachyonURI getPartitionUri(RawColumn column, int partitionId)
```
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/table/AbstractTachyonRawTables.java]: Uri or URI?

File: examples/src/main/java/tachyon/examples/BasicOperations.java [hunk 8]
```diff
@@ -44,18 +43,16 @@
   private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
   private final TachyonURI mMasterLocation;
   private final TachyonURI mFilePath;
-  private final InStreamOptions mReadOptions;
-  private final OutStreamOptions mWriteOptions;
+  private final OpenFileOptions mReadOptions;
```
Comment (jsimsa) [inline examples/src/main/java/tachyon/examples/BasicOperations.java]: mOpenFileOptions?

File: examples/src/main/java/tachyon/examples/BasicOperations.java [hunk 9]
```diff
@@ -44,18 +43,16 @@
...
+  private final CreateFileOptions mWriteOptions;
```
Comment (jsimsa) [inline examples/src/main/java/tachyon/examples/BasicOperations.java]: mCreateFileOptions?

**Comment by jsimsa** [comment]
@calvinjia overall this looks good ... just a couple of comments

File: clients/unshaded/src/main/java/tachyon/client/file/options/CreateFileOptions.java (hunk 5 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/options/CreateFileOptions.java]: Oh this is just to be consistent with our previous defaults for `create`

File: clients/unshaded/src/main/java/tachyon/client/file/FileSystem.java (hunk 6 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/FileSystem.java]: Good point, I've changed both to use `get()`. I changed the constructor behavior from a singleton to new instances since there currently is no state in the FileSystems.

**Comment by jsimsa** [comment]
LGTM

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java [hunk 10]
```diff
@@ -21,80 +21,69 @@
...
+   * @param tfs a {@link tachyon.client.file.FileSystem} handler
    * @param fileURI URI of the file
    * @param options client options to create the file with
    * @param len file size
-   * @return the {@link TachyonFile} of the created file
    * @throws IOException if {@code path} is invalid (e.g., illegal URI)
    */
-  public static TachyonFile createByteFile(TachyonFileSystem tfs, TachyonURI fileURI,
-      OutStreamOptions options, int len) throws IOException {
+  public static void createByteFile(FileSystem tfs, TachyonURI fileURI,
```
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: options last, like above

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java [hunk 11]
```diff
@@ -21,80 +21,69 @@
...
-import tachyon.client.file.options.OutStreamOptions;
-import tachyon.exception.ExceptionMessage;
+import tachyon.client.file.FileSystem;
+import tachyon.client.file.URIStatus;
+import tachyon.client.file.options.CreateFileOptions;
+import tachyon.client.file.options.OpenFileOptions;
 import tachyon.exception.TachyonException;
-import tachyon.thrift.FileInfo;
 
 public final class TachyonFSTestUtils {
```
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: Should this be renamed to `TachyonFileSystemTestUtils`?

File: clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java [hunk 12]
```diff
@@ -58,7 +58,7 @@
   protected FileSystemContext mContext;
 
   /**
-   * Constructor, currently {@link TachyonFileSystem} does not retain any state
+   * Constructor, currently {@link FileSystem} does not retain any state
```
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java]: Remove this comment? It seems like it should be in `FileSystem`, but it isn't very useful anyway.

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java [hunk 13]
```diff
@@ -102,30 +99,23 @@ public static boolean waitCompleted(TachyonFileSystemCore tfs, TachyonURI uri)
...
-        file = tfs.openIfExists(uri, OpenOptions.defaults());
-        if (file == null) {
-          LOG.debug("The file {} being waited upon does not exist yet. Waiting for it to be "
-              + "created.", uri);
-        }
-      }
-
-      if (file != null) {
-        completed = tfs.getInfo(file, GetInfoOptions.defaults()).isCompleted;
+      if (!tfs.exists(uri)) {
```
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java]: Beautiful :D

File: clients/unshaded/src/main/java/tachyon/client/table/TachyonRawTablesCore.java [hunk 14]
```diff
@@ -82,26 +82,26 @@ FileOutStream createPartition(RawColumn column, int partitionId)
   int getPartitionCount(RawColumn column) throws IOException, TachyonException;
 
   /**
-   * Gets a handle for the given raw table, if it exists.
-   *
-   * @param path the path of the table in Tachyon space
-   * @return a {@link RawTable} representing the table
+   * Gets the partition uri which can be used to access the partition data directly as a {@link
```
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/client/table/TachyonRawTablesCore.java]: uri -> URI

File: clients/unshaded/src/main/java/tachyon/client/table/TachyonRawTablesCore.java [hunk 15]
```diff
@@ -82,26 +82,26 @@ FileOutStream createPartition(RawColumn column, int partitionId)
...
+   * TachyonFile}
```
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/client/table/TachyonRawTablesCore.java]: space before params

File: clients/unshaded/src/main/java/tachyon/client/table/TachyonRawTablesCore.java [hunk 16]
```diff
@@ -82,26 +82,26 @@ FileOutStream createPartition(RawColumn column, int partitionId)
...
+   * @param column the column to get the partition from
+   * @param partitionId the id of the partition
+   * @return the uri referencing the partition
```
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/client/table/TachyonRawTablesCore.java]: uri -> URI

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 17]
```diff
@@ -306,30 +289,32 @@ public FileStatus getFileStatus(Path path) throws IOException {
     if (mStatistics != null) {
       mStatistics.incrementReadOps(1);
     }
-    FileInfo fileStatus;
+    URIStatus fileStatus;
     try {
-      TachyonFile file = mTFS.open(tPath);
-      fileStatus = mTFS.getInfo(file);
+      fileStatus = mTFS.getStatus(tPath);
     } catch (InvalidPathException e) {
```
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: InvalidPathException can't be thrown now

File: clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java [hunk 18]
```diff
@@ -70,40 +67,36 @@
   private int mBufferPosition = 0;
   private byte[] mBuffer;
 
-  private final TachyonConf mTachyonConf;
-
   /**
    * Constructs a new stream for reading a file from HDFS.
    *
-   * @param fileId the file id
+   * @param uri the Tachyon file uri
```
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: uri -> URI

File: examples/src/main/java/tachyon/examples/Performance.java [hunk 19]
```diff
@@ -68,10 +66,9 @@
   public static void createFiles() throws TachyonException, IOException {
     final long startTimeMs = CommonUtils.getCurrentMs();
     for (int k = 0; k < sFiles; k ++) {
-      TachyonFile file = sTFS.create(new TachyonURI(sFileName + (k + sBaseFileNumber)));
-      long fileId = file.getFileId();
+      sTFS.createFile(new TachyonURI(sFileName + (k + sBaseFileNumber))).close();
       LOG.info(
```
Comment (aaudiber) [inline examples/src/main/java/tachyon/examples/Performance.java]: this could go on one line now

File: integration-tests/src/test/java/tachyon/client/StreamOptionUtils.java [hunk 20]
```diff
@@ -15,84 +15,78 @@
...
+import tachyon.client.file.options.CreateFileOptions;
+import tachyon.client.file.options.OpenFileOptions;
 import tachyon.client.file.policy.LocalFirstPolicy;
 import tachyon.conf.TachyonConf;
 
 /**
  * A util class to obtain common In/OutStreamOptions for tests
- *
  */
+// TODO(calvin): We can make these methods into constants
```
Comment (aaudiber) [inline integration-tests/src/test/java/tachyon/client/StreamOptionUtils.java]: JIRA?

File: servers/src/main/java/tachyon/web/WebInterfaceWorkerBlockInfoServlet.java [hunk 21]
```diff
@@ -184,64 +182,54 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
...
-   * @param tachyonFileSystem the {@link TachyonFileSystem} client
+   * @param fileSystem the {@link tachyon.client.file.FileSystem} client
    * @param fileId the file id of the file
    * @param filePath the path of the file. valid iff fileId is -1
    * @return the {@link UIFileInfo} object of the file
    * @throws FileDoesNotExistException if the file does not exist
    * @throws IOException if an I/O error occurs
    */
-  private UIFileInfo getUiFileInfo(TachyonFileSystem tachyonFileSystem, long fileId,
+  private UIFileInfo getUiFileInfo(FileSystem fileSystem, long fileId,
```
Comment (aaudiber) [inline servers/src/main/java/tachyon/web/WebInterfaceWorkerBlockInfoServlet.java]: drop the `fileId` parameter?

File: shell/src/main/java/tachyon/shell/command/CatCommand.java [hunk 22]
```diff
@@ -37,7 +36,7 @@
    * @param conf the configuration for Tachyon
    * @param tfs the filesystem of Tachyon
    */
-  public CatCommand(TachyonConf conf, TachyonFileSystem tfs) {
+  public CatCommand(TachyonConf conf, FileSystem tfs) {
```
Comment (aaudiber) [inline shell/src/main/java/tachyon/shell/command/CatCommand.java]: tfs -> fs here and throughout this package

File: shell/src/main/java/tachyon/shell/command/LoadCommand.java [hunk 23]
```diff
@@ -63,25 +61,22 @@ void runCommand(TachyonURI path) throws IOException {
...
-      if (fInfo.isFolder) {
-        List<FileInfo> files = mTfs.listStatus(fd);
-        Collections.sort(files);
-        for (FileInfo file : files) {
-          TachyonURI newPath = new TachyonURI(file.getPath());
+      URIStatus status = mTfs.getStatus(filePath);
+      if (status.isFolder()) {
+        List<URIStatus> statuses = mTfs.listStatus(filePath);
+        for (URIStatus uriStatus : statuses) {
+          TachyonURI newPath = new TachyonURI(uriStatus.getPath());
```
Comment (aaudiber) [inline shell/src/main/java/tachyon/shell/command/LoadCommand.java]: why were we sorting before but aren't now?

File: shell/src/main/java/tachyon/shell/command/RmCommand.java [hunk 24]
```diff
@@ -44,24 +42,17 @@ public String getCommandName() {
...
-      throw new IOException(e.getMessage());
-    }
-
-    if (fInfo.isFolder) {
-      throw new IOException("rm: cannot remove a directory, please try rmr <path>");
-    }
-
-    try {
-      mTfs.delete(fd);
+      if (!mTfs.exists(path)) {
```
Comment (aaudiber) [inline shell/src/main/java/tachyon/shell/command/RmCommand.java]: Instead of checking state before calling delete, could we immediately call delete and then handle exceptions appropriately?

File: shell/src/main/java/tachyon/shell/command/UnmountCommand.java [hunk 25]
```diff
@@ -46,11 +46,8 @@ public void run(String... args) throws IOException {
...
     try {
-      if (mTfs.unmount(inputPath)) {
-        System.out.println("Unmounted " + inputPath);
-      } else {
-        throw new IOException("unmount: Failed to unmount" + inputPath);
-      }
+      mTfs.unmount(inputPath);
+      System.out.println("Unmounted " + inputPath);
     } catch (TachyonException e) {
       throw new IOException(e.getMessage());
```
Comment (aaudiber) [inline shell/src/main/java/tachyon/shell/command/UnmountCommand.java]: IOException(e)

**Comment by aaudiber** [comment]
Looks good overall, thanks for this great undertaking! I've left some comments

File: clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java (hunk 12 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java]: I think the class will be removed later.

File: clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java (hunk 18 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: the second `uri`?

File: shell/src/main/java/tachyon/shell/command/CatCommand.java (hunk 22 above)
Comment (calvinjia) [inline shell/src/main/java/tachyon/shell/command/CatCommand.java]: will do the renaming separately.

File: shell/src/main/java/tachyon/shell/command/LoadCommand.java (hunk 23 above)
Comment (calvinjia) [inline shell/src/main/java/tachyon/shell/command/LoadCommand.java]: I'm not sure what the point of sorting was, since `FileInfo`'s comparison isn't really useful.

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java (hunk 10 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: Do you mind having `len` last? It seems like the other methods all have `len` last.

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java (hunk 11 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: Will do renaming in a separate PR.

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 17 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Updated, thanks.

File: examples/src/main/java/tachyon/examples/Performance.java (hunk 19 above)
Comment (calvinjia) [inline examples/src/main/java/tachyon/examples/Performance.java]: Updated, thanks.

File: integration-tests/src/test/java/tachyon/client/StreamOptionUtils.java (hunk 20 above)
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/client/StreamOptionUtils.java]: Created: https://tachyon.atlassian.net/browse/TACHYON-1524

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java (hunk 10 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: Options is sort of like an extension point to add new parameters without changing the signature, so it feel like it belongs at the end, similar to varargs or the FileSystem API

File: clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java (hunk 18 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: yea, it's an acronym

File: integration-tests/src/test/java/tachyon/client/StreamOptionUtils.java (hunk 20 above)
Comment (aaudiber) [inline integration-tests/src/test/java/tachyon/client/StreamOptionUtils.java]: thanks

File: shell/src/main/java/tachyon/shell/command/LoadCommand.java (hunk 23 above)
Comment (aaudiber) [inline shell/src/main/java/tachyon/shell/command/LoadCommand.java]: sgtm

File: shell/src/main/java/tachyon/shell/command/CatCommand.java (hunk 22 above)
Comment (aaudiber) [inline shell/src/main/java/tachyon/shell/command/CatCommand.java]: seems like it belongs in this PR since we rename tfs to fs elsewhere, and the type is changing, but I can wait if you want to get this in ASAP

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java (hunk 10 above)
Comment (uronce-cc) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: +1, good to keep the parameters ordered in the same fashion.

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java [hunk 26]
```diff
@@ -160,15 +138,17 @@ public static TachyonFile createByteFile(TachyonFileSystem tfs, String fileName,
   }
 
   /**
-   * Converts an {@link OutStreamOptions} object to an {@link InStreamOptions} object with a
+   * Converts an {@link CreateFileOptions} object to an {@link OpenFileOptions} object with a
```
Comment (uronce-cc) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: NIT: `a {@link Create...}`

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java [hunk 27]
```diff
@@ -160,15 +138,17 @@ public static TachyonFile createByteFile(TachyonFileSystem tfs, String fileName,
...
    * matching Tachyon storage type.
    *
-   * @param op an {@link OutStreamOptions} object
-   * @return an {@link InStreamOptions} object with a matching Tachyon storage type
+   * @param op an {@link CreateFileOptions} object
```
Comment (uronce-cc) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: NIT: `a {@link Create...`

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java (hunk 13 above)
Comment (uronce-cc) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java]: Finally, we have the `exists` API :)

File: examples/src/main/java/tachyon/examples/BasicNonByteBufferOperations.java [hunk 28]
```diff
@@ -93,33 +89,25 @@ private void write(TachyonFileSystem tachyonClient) throws IOException, TachyonE
...
-      boolean deleteIfExists, OutStreamOptions clientOptions) throws IOException, TachyonException {
-    TachyonFile file;
-
-    try {
-      file = tachyonFileSystem.open(filePath);
-    } catch (Exception e) {
-      file = null;
-    }
-    if (file == null) {
+  private FileOutStream getOrCreate(FileSystem fileSystem, TachyonURI filePath,
```
Comment (uronce-cc) [inline examples/src/main/java/tachyon/examples/BasicNonByteBufferOperations.java]: Why this is called `getOrCreate`? At first sight, I thought this means "if a file exits, get outstream for that file, otherwise, create a file at path, and get its outstream."

**Comment by uronce-cc** [comment]
LGTM, just a few minor comments.

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java [hunk 29]
```diff
@@ -180,7 +169,7 @@ public static long persistFile(TachyonFileSystem tfs, TachyonFile file, FileInfo
       closer.close();
     }
     // Tell the master to mark the file as persisted
-    tfs.setState(file, (new SetStateOptions.Builder()).setPersisted(true).build());
+    fs.setAttribute(uri, SetAttributeOptions.defaults().setPersisted(true));
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java]: just for curiosity, does this overwrite the existing attributes on master (except setting persisted true)?

File: examples/src/main/java/tachyon/examples/BasicNonByteBufferOperations.java [hunk 30]
```diff
@@ -72,16 +71,13 @@ public Boolean call() throws Exception {
...
     write(tachyonClient);
     return read(tachyonClient);
   }
 
-  private void write(TachyonFileSystem tachyonClient) throws IOException, TachyonException {
-    OutStreamOptions clientOptions =
-        new OutStreamOptions.Builder(ClientContext.getConf()).setWriteType(mWriteType).build();
-    FileOutStream fileOutStream =
-        getOrCreate(tachyonClient, mFilePath, mDeleteIfExists, clientOptions);
+  private void write(FileSystem tachyonClient) throws IOException, TachyonException {
```
Comment (apc999) [inline examples/src/main/java/tachyon/examples/BasicNonByteBufferOperations.java]: `tachyonClient` -> `fileSystem'?

File: integration-tests/src/test/java/tachyon/client/FileSystemIntegrationTest.java [hunk 31]
```diff
@@ -124,22 +122,21 @@ public void getFileStatusTest() throws IOException, TachyonException {
...
-    Assert.assertNotNull(fileInfo);
-    Assert.assertTrue(fileInfo.getPath().equals(uniqPath));
+    TachyonFSTestUtils.createByteFile(mTfs, uri.getPath(), writeBytes, mWriteBoth);
+    Assert.assertTrue(mTfs.getStatus(uri).getInMemoryPercentage() == 100);
+
+    Assert.assertTrue(mTfs.getStatus(uri).getPath().equals(uniqPath));
   }
 
   @Test
   public void mkdirTest() throws IOException, TachyonException {
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/client/FileSystemIntegrationTest.java]: shall we call this `createDirectoryTest`?

File: integration-tests/src/test/java/tachyon/hadoop/TFSStatisticsTest.java [hunk 32]
```diff
@@ -44,21 +42,21 @@
   @ClassRule
   public static LocalTachyonClusterResource sLocalTachyonClusterResource =
       new LocalTachyonClusterResource(10000, 1000, BLOCK_SIZE);
-  private static FileSystem.Statistics sStatistics;
-  private static FileSystem sTFS;
+  private static org.apache.hadoop.fs.FileSystem.Statistics sStatistics;
+  private static org.apache.hadoop.fs.FileSystem sTFS;
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/hadoop/TFSStatisticsTest.java]: why this is called `sTFS`?

File: integration-tests/src/test/java/tachyon/master/MasterFaultToleranceIntegrationTest.java [hunk 33]
```diff
@@ -143,9 +142,9 @@ public void deleteFileFaultTest() throws Exception {
 
         // We can not call mTfs.delete(mTfs.open(new TachyonURI(TachyonURI.SEPARATOR))) because root
         // node can not be deleted.
-        for (FileInfo file : mTfs.listStatus(mTfs.open(new TachyonURI(TachyonURI.SEPARATOR)))) {
-          mTfs.delete(new TachyonFile(file.getFileId()),
-              new DeleteOptions.Builder().setRecursive(true).build());
+        for (URIStatus file : mTfs.listStatus(new TachyonURI(TachyonURI.SEPARATOR))) {
+          mTfs.delete(new TachyonURI(file.getPath()), new DeleteOptions.Builder()
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/master/MasterFaultToleranceIntegrationTest.java]: why `DeleteOptions` is still using the `Builder()` model?

File: shell/src/main/java/tachyon/shell/command/UnmountCommand.java (hunk 25 above)
Comment (apc999) [inline shell/src/main/java/tachyon/shell/command/UnmountCommand.java]: there are many cases like this, maybe this should be done by a separate PR.

**Comment by apc999** [comment]
@calvinjia  finished my pass, left a few minor comments.

File: integration-tests/src/test/java/tachyon/master/MasterFaultToleranceIntegrationTest.java (hunk 33 above)
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/master/MasterFaultToleranceIntegrationTest.java]: I plan to change all the options in a following commit.

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java (hunk 29 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemUtils.java]: The defaults for `SetAttributeOptions` are unset attributes, so it will only update setting persisted to true and retain the previous values for others.

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java (hunk 26 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: Updated, thanks.

File: clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java (hunk 27 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFSTestUtils.java]: Updated, thanks.

File: examples/src/main/java/tachyon/examples/BasicNonByteBufferOperations.java (hunk 28 above)
Comment (calvinjia) [inline examples/src/main/java/tachyon/examples/BasicNonByteBufferOperations.java]: I've updated the name to `createFile`.

File: clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java (hunk 18 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: Makes sense, updated.

File: servers/src/main/java/tachyon/web/WebInterfaceWorkerBlockInfoServlet.java (hunk 21 above)
Comment (calvinjia) [inline servers/src/main/java/tachyon/web/WebInterfaceWorkerBlockInfoServlet.java]: Updated, currently its a very basic solution, I've opened a ticket for a more long term solution. https://tachyon.atlassian.net/browse/TACHYON-1525

File: shell/src/main/java/tachyon/shell/command/RmCommand.java (hunk 24 above)
Comment (calvinjia) [inline shell/src/main/java/tachyon/shell/command/RmCommand.java]: Good idea, I left a TODO here because the exception messages don't make too much sense yet, since we haven't moved completely to communicating with TachyonURIs with the master.

File: shell/src/main/java/tachyon/shell/command/CatCommand.java (hunk 22 above)
Comment (calvinjia) [inline shell/src/main/java/tachyon/shell/command/CatCommand.java]: Would prefer to get the logic reviewed and then follow up with the renaming.

File: integration-tests/src/test/java/tachyon/hadoop/TFSStatisticsTest.java (hunk 32 above)
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/hadoop/TFSStatisticsTest.java]: It is a `TFS` object (implements hadoop FileSystem)

File: integration-tests/src/test/java/tachyon/client/FileSystemIntegrationTest.java (hunk 31 above)
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/client/FileSystemIntegrationTest.java]: Will rename after the logical changes are done.

File: examples/src/main/java/tachyon/examples/BasicNonByteBufferOperations.java (hunk 30 above)
Comment (calvinjia) [inline examples/src/main/java/tachyon/examples/BasicNonByteBufferOperations.java]: Will rename after all the logical changes are done. As a side note, do you prefer `fileSystem` or `fs`?

Comment (jsimsa) [inline examples/src/main/java/tachyon/examples/BasicNonByteBufferOperations.java]: FWIW, if it is `fileSystem` then IDEs such as IntelliJ will be smart enough do a refactor.

**Comment by calvinjia** [comment]
@apc999 @uronce-cc @aaudiber 
Comments addressed.
I plan to have another PR for the options changes, master side changes, and finally renaming.

**Comment by aaudiber** [comment]
LGTM

File: examples/src/main/java/tachyon/examples/BasicNonByteBufferOperations.java (hunk 30 above)
Comment (jsimsa) [inline examples/src/main/java/tachyon/examples/BasicNonByteBufferOperations.java]: That is next time we are renaming this class from `FileSystem` to `TFS`

**Comment by jsimsa** [comment]
LGTM

**Comment by apc999** [comment]
+1 LGTM
